(GH-2140) Document escalating privilege in powershell scripts on Windows

lucywyman opened the pull request:
The WinRM transport does not - and will not - support something akin to
a `run-as` command. This is because the way WinRM works is by connecting
with the user that has permissions to perform the actions, not by
elevating or changing to another user after connection. Put another way,
WinRM operates with a session level token and has no functionality to
change to another user after authentication.

Instead of escalating the user once connected, users can pass
credentials to scripts or plans running on the targets and use them
directly in commands. There are a number of ways to do this, each ideal
for different use cases. This adds a section to the 'Escalating
Privilege' document that describes the different ways to pass
credentials securely to either a script or plan to be used for
escalating on Windows. This includes using sensitive task parameters,
and describes setting and retrieving variables on Target objects
securely using Hiera data, Inventory plugins, and Remote tasks.

Closes #2140

!no-release-notes

lucywyman commented:
I wasn't entirely sure of what examples to provide and where, since most of these things aren't specific to windows and the workflows are so distinct. I don't think writing an example for each possible workflow would be a good use of time, and there are already smaller (platform-agnostic) examples of each element of the workflow on their respective pages. I tested each workflow on a Windows machine to verify it worked as expected, so happy to include examples anywhere they would fit!

beechtom reviewed (changes requested):
@lucywyman This is a generated doc, so the changes should be in the template.

lucywyman commented:
Booooo updated

puppetcla commented:
CLA signed by all contributors.

hestonhoffman commented on documentation/templates/privilege_escalation.md.erb:
```diff
@@ -99,3 +99,63 @@ This enables using any executable and command-line options to change users. Keep
+
+## Escalating privilege over WinRM
+
+As you may have noted we do not support `run-as` for the WinRM transport. This is because the way
+WinRM works is by connecting with the user that has permissions to perform the actions, not by
+elevating or changing to another user after connection. WinRM operates with a session level token
+and has no functionality to change to another user after authentication. Instead, you can connect
+with one user, then provide credentials to a program that accepts them to perform another action,
+but that is not part of WinRm.
```
```suggestion
The WinRM transport does not support `run-as`. This is because
WinRM operates with a session level token and has no functionality to change to another user after
authentication. Once you connect with a user, you can only perform actions that you have permission to
perform with that user.  Instead of switching users, you can connect with a user, and then provide credentials
to a program that accepts them to perform another action. The behavior of providing credentials to a 
program is not part of WinRM.
```

hestonhoffman commented on documentation/templates/privilege_escalation.md.erb:
```diff
@@ -99,3 +99,63 @@ This enables using any executable and command-line options to change users. Keep
+with one user, then provide credentials to a program that accepts them to perform another action,
+but that is not part of WinRm.
+
+For example, you can use `Invoke-BoltScript -Script .\myscript.ps1 -Targets winrm://mytarget -User
+userA -Password passwordA` to connect as a user you have privilege to connect as, then within
+`myscript.ps1` you can run:
```
```suggestion
For example, to get the processes running on a Windows target, you could connect to the
target with WinRM using 
`Invoke-BoltScript -Script .\myscript.ps1 -Targets winrm://mytarget -User userA -Password passwordA`
and then use the following Powershell script to pass credentials to the `Get-WmiObject` command:
```

hestonhoffman reviewed (commented):
I have a couple comments, but I've run out of time! I'll get back to this on Monday :-)

hestonhoffman commented on documentation/templates/privilege_escalation.md.erb:
```diff
@@ -99,3 +99,65 @@ This enables using any executable and command-line options to change users. Keep
+parameters](writing_tasks.md#defining-sensitive-parameters) for that task. This will obfuscate the
+secret in logs and API responses.
+
+You can pass sensitive parameters to a task directly if your secrets are the same for every target.
```
Does "directly" here mean via the command line?

lucywyman replied:
Yeah, at the time I was thinking "without having to do anything fancy involving plans or inventory", but I think 'via the command line or directly to `run_task`' would also work?

hestonhoffman commented on documentation/templates/privilege_escalation.md.erb:
```diff
@@ -99,3 +99,65 @@ This enables using any executable and command-line options to change users. Keep
+
+You can pass sensitive parameters to a task directly if your secrets are the same for every target.
+If your secrets are unique to each target, or need to be computed, fetched, or obfuscated when you
+pass them to the task you will want to set your secret as data on the target in the inventory. For
```
I think a small inventory example here would be really helpful. Would you set this under the `vars` key?

hestonhoffman commented on documentation/templates/privilege_escalation.md.erb:
```diff
@@ -99,3 +99,65 @@ This enables using any executable and command-line options to change users. Keep
+the purposes of escalating as a secondary user you can set arbitrary variables on targets in the
+inventory, which you can then access.
+
+There are a few ways of setting and retrieving variables on targets. You can set variables on a
```
I'm a little confused here. I think it's just cos there's a lot going on in this paragraph and I'm not sure if it's describing an alternate way to do the same thing as the previous paragraph, or an additional thing that you need to do. I'm guessing it's an alternate way based on the summary later.

So far, we've covered using a task, defining sensitive parameters, and passing sensitive parameters into the task using an inventory file or directly. Is this paragraph covering a new use case - using an inventory plugin or Hiera data along with a remote task instead of defining sensitive parameters?

lucywyman replied:
I'm just going to remove the Hiera bit. You can definitely use it along these lines, but it's so complicated and I think anyone who isn't familiar with Hiera can just do the inventory workflow

hestonhoffman commented on documentation/templates/privilege_escalation.md.erb:
```diff
@@ -99,3 +99,65 @@ This enables using any executable and command-line options to change users. Keep
+have to manage the connection to the remote target in your task rather than letting Bolt manage it
+for you.
+
+To summarize:
```
Instead of summarizing at the end, let's move the summary bit to the top - that's a good way to signpost to the reader that you're going to discuss two possible options/use cases

lucywyman commented:
Ok, I rewrote this to be more clear and tried to make it more clear how to choose which solution was best. I removed the Hiera bit because it's a whole other can of worms and I think this solution is simpler and will work for most users, but using Hiera is really powerful and kind of cool.

I feel like I said the same thing like 3 times, could definitely use a good editor!

hestonhoffman commented:
Hi Lucy, I was having trouble editing this in the GUI, so I pushed up a PR to your fork instead: https://github.com/lucywyman/bolt/pull/11

lucywyman commented:
Merged and made some small edits

hestonhoffman reviewed (approved):
LGTM!
